terminal/src/main/scala/ammonite/terminal/filters/ReadlineFilters.scala

```diff
@@ -120,13 +120,53 @@ object ReadlineFilters {
       (b patch(from = c - 1, patch = Nil, replaced = 1), c - 1)
     }
 
-    def cutAllLeft(b: Vector[Char], c: Int) = {
-      prepend(b.take(c))
-      (b.drop(c), 0)
+    def cutLineLeft(b: Vector[Char], c: Int) = {
+      val (allBeforeCursor, allAfterCursor) = b.splitAt(c)
+      val previousNewlineIndex = allBeforeCursor.lastIndexWhere(_ == '\n')
+      if (previousNewlineIndex == -1) {
+        // beginning of input. no leading newline
+        prepend(allBeforeCursor)
+        (allAfterCursor, 0)
+      } else {
+        val (allBeforeLine, lineBeforeCursor) = allBeforeCursor.splitAt(
+          previousNewlineIndex
+        )
+        prepend(lineBeforeCursor)
+        val charsBeforeCursorOnLine = lineBeforeCursor.length
+        if (charsBeforeCursorOnLine == 1) {
+          // if only a newline before cursor on line, cut it
+          (allBeforeLine ++ allAfterCursor, c - 1)
+        } else {
+          // if there's more on the line before cursor, cut to beginning of line
+          (
+            allBeforeLine ++ "\n" ++ allAfterCursor,
```

Comment:
Let's assign these to temporary `val`s and construct the tuple from those, rather than having these somewhat-awkward multi-line tuples